src/iterators.rs

```diff
@@ -0,0 +1,81 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use std::convert::TryInto;
+
+use pyo3::class::iter::{IterNextOutput, PyIterProtocol};
+use pyo3::class::PySequenceProtocol;
+use pyo3::exceptions::PyIndexError;
+use pyo3::prelude::*;
+
+/// A custom iterator class for the return from :func:`retworkx.bfs_successors`
+///
+/// This class is a container class for the results of the
+/// :func:`retworkx.bfs_successors` function. It implements both the Python
+/// iterator protocol and sequence protocol. So you can treat the return as
+/// either a read-only sequence/list that is integer indexed or use it as an
+/// iterator that will yield the results in order.
+///
+/// For example::
+///
+///     import retworkx
+///
+///     graph = retworkx.generators.directed_path_graph(5)
+///     bfs_succ = retworkx.bfs_successors(0)
+///     # Index based access
+///     third_element = bfs_succ[2]
+///     # Use as iterator
+///     first_element = next(bfs_succ)
+///     second_element = nex(bfs_succ)
+///
+#[pyclass(module = "retworkx")]
+pub struct BFSSuccessors {
+    pub bfs_successors: Vec<(PyObject, Vec<PyObject>)>,
+    pub index: usize,
+}
+
+#[pyproto]
+impl PySequenceProtocol for BFSSuccessors {
+    fn __len__(&self) -> PyResult<usize> {
+        Ok(self.bfs_successors.len())
+    }
+
+    fn __getitem__(
+        &'p self,
+        idx: isize,
+    ) -> PyResult<(PyObject, Vec<PyObject>)> {
+        if idx >= self.bfs_successors.len().try_into().unwrap() {
```

Comment:
Would it make sense to check that `idx` is non-negative here? (Is `isize` signed?)

Reply:
Oh, actually this is a good point. This will totally fall apart if you pass a negative number in. I think I need to add a bit more logic here to handle negatives, because I assume the expectation from users is if you were do `bfs_successors(graph)[-1]` you'd get the last entry, which definitely won't work like this.

Also yes [`isize`](https://doc.rust-lang.org/std/primitive.isize.html) is a pointer sized signed integer. (compared to [`usize`](https://doc.rust-lang.org/std/primitive.usize.html) which is an unsigned pointer sized integer

Reply:
So I actually tested this with `bfs_successors(graph)[-1]` and it works as expected (which I'm surprised about because indices for Vec are unsigned) so I think it's something python is doing for us. I added a test to assert negative indices work as expected. So I don't think we need to change anything here. What do you think?

Reply:
:+1: Agree, the python behavior looks good. Thanks for adding the test.